### PR TITLE
Nobid Prebid Adapter commit

### PR DESCRIPTION
**Pull request by robdubois**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

Initial pull request for Nobid prebid bid adapter.

- test parameters for validating bids
```
{
  bidder: 'nobid',
  params: {
    siteId: SITE_ID_PROVIDED_BY_YOUR_NOBID_ACCOUNT_MANAGER
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- rob.dubois@nobid.io
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- Link to a PR on the docs repo https://github.com/prebid/prebid.github.io/pull/1416

### CODE REVIEW

**Comment by mkendall07** [comment]
This pull request **introduces 1 alert** when merging 67798a5107143bea9bc140ac66de27175e5cfaae into 9ed95bb067ffcf457c90be9c716df5cf71bc2a9d - [view on LGTM.com](https://lgtm.com/projects/g/prebid/Prebid.js/rev/pr-2db4828e4e36a8dc47abf39e398886ea340205f9)

**new alerts:**

* 1 for Incomplete string escaping or encoding

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by robdubois** [comment]
circleci shows an error but when I go the logs I am not seeing any errors.
Do you what the problem is?

**Comment by robdubois** [comment]
Changes have been applied. Please review. Thank you.

File: modules/nobidBidAdapter.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            mediaTypes: {
+                banner: {
+                    sizes: [[320, 50]],   // a mobile size
+                }
+            },
+            bids: [
+                {
+                    bidder: "nobid",
+                    params: {
+                        siteId: <SITE_ID_PROVIDED_BY_YOUR_NOBID_ACCOUNT_MANAGER>
```
Comment (robertrmartinez) [inline modules/nobidBidAdapter.md]: Thanks for submitting this.

However every bidders markdown is required to have valid Test ad units which will consistently return valid bids in order to be tested by PR reviewers as well as users.

Please see the [Required Files](http://prebid.org/dev-docs/bidder-adaptor.html#required-files) Section in the Docs for more information on the requirements on adding a New Bid Adapter.

**Comment by robdubois** [comment]
Hello,
Can you please review the documentation. 
Thank you

**Comment by robertrmartinez** [comment]
@robdubois Yeah I am reviewing.

It is unusual for an adapter to create an object on the window as you seem to be doing.

Can you elaborate why this is?

One of the things we look for is for adapters to not rely on the creation and maintaining of global objects in the adapter.

Can you explain why this is needed?

Also there are several instances where you are using your own log and even the regular console.log.

Is there a reason why you cannot just utilize the prebid utils logger?
Then, you won't need to check for a "nobid" debug flag and it will work the same way all the other adapters logging works, using the debug = true in setConfig or pbjs_debug=true in the query param.

I will be doing some extra testing of different auction scenarios to ensure this global nobid object works as needed. But still want to understand why the adapter cannot use the already built prebid adapter specs instead of adding this as well.

Also, not sure if it is related to your PR, but I cannot get your branch to pass on Circle CI at all. Usually these are resolved after a rebuild, but I have rebuilt many times already.

I will investigate and find out what is wrong.

**Comment by robertrmartinez** [comment]
@robdubois 

Also there was a package updated in master which may fix the Circle CI tests failing, please merge master into your branch when you have a chance!

**Comment by robdubois** [comment]
@robertrmartinez ,
Thanks for the feedback. Having an object is cleaner design it also helps keeping state. I think I can easily change it to stateless to avoid having issues if the Adapter is present multiple times on the page. Let me change that.
I will make the necessary changes to the logger. Good point.

**Comment by robertrmartinez** [comment]
@robdubois Awesome sounds perfect. Appreciate it!

Hopefully after your changes and pulling the latest from master we can get the tests to pass and get this thing merged!

**Comment by robdubois** [comment]
@robertrmartinez ,
I made the changes to the adapter. We still need a stripped down version of the nobid object because our adapter has a passback to a DFP account and needs access to the nobid.renderTag function.

**Comment by robdubois** [comment]
@robertrmartinez ,
It should be good now. All ad markups have an id in map. I also merged the upstream master.

**Comment by robertrmartinez** [comment]
Awesome stuff. Tests are finally passing!

Code looks good, Just want to do some testing using a few test pages to make sure. Thanks

**Comment by robdubois** [comment]
Great. Let me know. Thanks

File: modules/nobidBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  },
+
+  /**
+     * Register the user sync pixels which should be dropped after the auction.
+     *
+     * @param {SyncOptions} syncOptions Which user syncs are allowed?
+     * @param {ServerResponse[]} serverResponses List of server's responses.
+     * @return {UserSync[]} The user syncs which should be dropped.
+     */
+  getUserSyncs: function(syncOptions, serverResponses) {
```
Comment (robertrmartinez) [inline modules/nobidBidAdapter.js]: FYI:

GDPR information can be passed into the getUserSyncs function and passed along to your sync endpoint.

See https://github.com/prebid/Prebid.js/blob/master/modules/rubiconBidAdapter.js#L602 as an example.

Maybe add it in a future PR if needed?

File: test/spec/modules/nobidBidAdapter_spec.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
+import { expect } from 'chai';
```
Comment (robertrmartinez) [inline test/spec/modules/nobidBidAdapter_spec.js]: Your test coverage falls below the 80% test coverage minimum requirement.

image

I think if you add a couple tests around `getUserSyncs` it should get above 80%.

If you have the time maybe try to cover some other lines/functions as well.

You can generate and view your test coverage by running the following:

```
gulp test-coverage
gulp view-coverage
```

We are cutting a new Prebid release tomorrow (Wednesday September 4th) around 1:00 PM PST.

If you get the test coverage up before then I will make sure to get it in the release of PBJS `2.31.0`!

**Comment by robdubois** [comment]
@robertrmartinez ,
Thanks for the review.
- added GDPR support in user sync
- The test coverage is now above 80%

File: modules/nobidBidAdapter.js (hunk 2 above)
Comment (robdubois) [reply modules/nobidBidAdapter.js]: This is done

File: test/spec/modules/nobidBidAdapter_spec.js (hunk 3 above)
Comment (robdubois) [reply test/spec/modules/nobidBidAdapter_spec.js]: Done too.

**Comment by robdubois** [comment]
Business is impatiently waiting for this release. I would love to see live Wednesday.
Thanks

**Pull request merged**